File: src/main/java/org/hwabeag/hwaskyblock/events/MoveEvent.java
package org.hwabeag.hwaskyblock.events;

import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;
import org.hwabeag.hwaskyblock.config.ConfigManager;

import java.util.Objects;

public class MoveEvent implements Listener {

    FileConfiguration Config = ConfigManager.getConfig("setting");
    FileConfiguration MessageConfig = ConfigManager.getConfig("message");
    FileConfiguration SkyBlockConfig = ConfigManager.getConfig("skyblock");
    String Prefix = ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("hwaskyblock-system.prefix")));

    @EventHandler
    public void onPlayerMove(PlayerMoveEvent event) {
        Player player = event.getPlayer();
        String name = player.getName();
        Location from = event.getFrom();
        Location to = event.getTo();
        Block blockFrom = from.getWorld().getBlockAt(from);
        Block blockTo = to.getWorld().getBlockAt(to);
        World world = blockTo.getWorld();
        String world_name = world.getWorldFolder().getName();
        String[] number = world_name.split("\\.");
        if (Objects.equals(number[0], "HwaSkyBlock")) {
            String block_to_id = number[1];
            if (SkyBlockConfig.getString(block_to_id + ".leader") != null) {
                if (!Objects.equals(SkyBlockConfig.getString(block_to_id + ".leader"), name)) {
                    if (SkyBlockConfig.getString(block_to_id + ".sharer." + name) == null) {
                        if (!SkyBlockConfig.getBoolean(block_to_id + ".join")) {
                            player.teleport(blockFrom.getLocation().add(0, 1, 0));
                            player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(MessageConfig.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                        }
                    } else {
                        if (!SkyBlockConfig.getBoolean(block_to_id + ".sharer." + name + ".join")) {
                            player.teleport(blockFrom.getLocation().add(0, 1, 0));
                            player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(MessageConfig.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                        }
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/events/player/MoveEvent.kt
package org.hwabaeg.hwaskyblock.events.player

import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent
import org.hwabaeg.hwaskyblock.database.DatabaseManager
import org.hwabaeg.hwaskyblock.database.config.ConfigManager
import java.util.*

class MoveEvent : Listener {
    var Config: FileConfiguration = ConfigManager.getConfig("setting")!!
    var MessageConfig: FileConfiguration = ConfigManager.getConfig("message")!!
    var Prefix: String = ChatColor.translateAlternateColorCodes(
        '&',
        Objects.requireNonNull<String?>(Config.getString("hwaskyblock-system.prefix"))
    )

    @EventHandler
    fun onPlayerMove(event: PlayerMoveEvent) {
        val player = event.getPlayer()
        val name = player.name
        val from = event.from
        val to = event.to
        val blockFrom = from.world?.getBlockAt(from)
        val blockTo = to?.world?.getBlockAt(to)
        val world = blockTo?.world
        val world_name = world?.worldFolder?.getName()
        val number: Array<String?> = world_name?.split("\\.".toRegex())?.dropLastWhile { it.isEmpty() }!!.toTypedArray()
        if (number[0] == "HwaSkyBlock") {
            val block_to_id = number[1]
            val leader = DatabaseManager.getSkyBlockData(
                block_to_id.toString(),
                "$block_to_id.leader",
                "getSkyBlockLeader"
            ) as? String
            if (leader != null && leader != name) {
                val isSharer = (DatabaseManager.getShareData(
                    block_to_id.toString(),
                    "",
                    "getShareList",
                    null
                ) as? List<*>)?.contains(name) == true
                val hasJoinPermission = if (isSharer) {
                    DatabaseManager.getShareData(block_to_id.toString(), name, "can_join", "isUseJoin") as? Boolean
                        ?: true
                } else {
                    DatabaseManager.getSkyBlockData(
                        block_to_id.toString(),
                        "$block_to_id.join",
                        "isSkyBlockJoin"
                    ) as? Boolean ?: true
                }
                if (!hasJoinPermission) {
                    blockFrom?.let {
                        player.teleport(it.location.add(0.0, 1.0, 0.0))
                    }
                    val message = ChatColor.translateAlternateColorCodes(
                        '&',
                        Prefix + Objects.requireNonNull(MessageConfig.getString("message-event.no_permission"))
                    )
                    player.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent(message))
                    event.isCancelled = true
                }
            }
        }
    }
}